### ISSUE TITLE
AKS+KIC+CIS - no pool members

### ISSUE THREAD

**Issue by nergalex** [F5Networks/k8s-bigip-ctlr#1579]
#### Setup Details
 CIS Version : 2.2.0
 Build: f5networks/k8s-bigip-ctlr:latest       
 BIGIP Version: Big IP 15.1.0.2
 AS3 Version: 3.24.0.5
 Agent Mode: AS3
 Orchestration: K8S
 Orchestration Version: 1.18.10
 Pool Mode: Nodeport      
 Additional Setup details:
- AKS
- CNI Azure

#### Description  
 Objective: deploy NGINX Ingress Controller + CIS
 Symptom: Partition, Virtual Server, pool are well created but pool does not have pool members.

#### Steps To Reproduce
1) Deploy NGINX IC with HELM
2) Deploy CIS with HELM
3) Deploy CIS ConfigMap
4) Change labels in NGINX IC's Service resource to as needed for CIS

#### Expected Result 
Pool contains pool members = K8S nodes + specified `nodeport`

#### Actual Result  
Pool does NOT contain pool members 

#### Diagnostic Information
[controller-logs.txt](https://github.com/F5Networks/k8s-bigip-ctlr/files/5584038/controller-logs.txt)
[controller-parameters.txt](https://github.com/F5Networks/k8s-bigip-ctlr/files/5584039/controller-parameters.txt)
[Deployment_f5-cis-big-ip-controller-f5-bigip-ctlr.yaml.TXT](https://github.com/F5Networks/k8s-bigip-ctlr/files/5584040/Deployment_f5-cis-big-ip-controller-f5-bigip-ctlr.yaml.TXT)
[Deployment-spoke-ingress-controller-standard.yaml.TXT](https://github.com/F5Networks/k8s-bigip-ctlr/files/5584041/Deployment-spoke-ingress-controller-standard.yaml.TXT)
[Service-spoke-ingress-controller-standard.yaml.TXT](https://github.com/F5Networks/k8s-bigip-ctlr/files/5584042/Service-spoke-ingress-controller-standard.yaml.TXT)
[ClusterRole_f5-cis-big-ip-controller-f5-bigip-ctlr.log](https://github.com/F5Networks/k8s-bigip-ctlr/files/5584043/ClusterRole_f5-cis-big-ip-controller-f5-bigip-ctlr.log)
[ConfigMap.yaml.TXT](https://github.com/F5Networks/k8s-bigip-ctlr/files/5584044/ConfigMap.yaml.TXT)

#### Observations (if any)
I'm sure it is a misconfiguration from my side :)
I tried also `ClusterIP `type... same issue

**Comment by mdditt2000** [comment]
@nergalex are you using NGINX Plus?

**Comment by nergalex** [comment]
Yes NGINX+

**Comment by mdditt2000** [comment]
@nergalex yes have this new feature called NCC "NginxCISConnector" which will solve of the complexity. It's still in preview. CIS using NCC is aware of NGINX IC. All you need todo is create the CRD in CIS.

https://github.com/F5Networks/k8s-bigip-ctlr/tree/master/docs/_static/config_examples/crd/NginxCisConnector

You will require a CIS + NGINX image. Let me know if you interested. If not we can look at you service to determine why service discovery is not working.

**Comment by nergalex** [comment]
OK thank you, I'll try

**Comment by nergalex** [comment]
Seems that current version of CIS does not process a NginxCisConnector resource.

[ClusterRoleBinding_f5-cis-big-ip-controller-f5-bigip-ctlr.yaml.TXT](https://github.com/F5Networks/k8s-bigip-ctlr/files/5586135/ClusterRoleBinding_f5-cis-big-ip-controller-f5-bigip-ctlr.yaml.TXT)
[ClusterRole_f5-cis-big-ip-controller-f5-bigip-ctlr.yaml.TXT](https://github.com/F5Networks/k8s-bigip-ctlr/files/5586130/ClusterRole_f5-cis-big-ip-controller-f5-bigip-ctlr.yaml.TXT)
[CustomResourceDefinition_nginxcisconnectors.cis.f5.com.yaml.TXT](https://github.com/F5Networks/k8s-bigip-ctlr/files/5586132/CustomResourceDefinition_nginxcisconnectors.cis.f5.com.yaml.TXT)
[NginxCisConnector_ncc-spoke-ingress-controller-standard.yaml.TXT](https://github.com/F5Networks/k8s-bigip-ctlr/files/5586133/NginxCisConnector_ncc-spoke-ingress-controller-standard.yaml.TXT)
[cis.log](https://github.com/F5Networks/k8s-bigip-ctlr/files/5586134/cis.log)

**Comment by mdditt2000** [comment]
@nergalex waiting for the image on the NGINX side. 

In the meantime let's get your ConfigMap working. 

1) ConfigMap

- If you want to use Configmap you need to create two ConfigMaps. One for port 80 and 443. In your CIS code the ConfigMap cannot detect both ports defined in the service. Since you Label are correct, i believe CIS is only picking up first port 80 and ignoring 443. This was fixed last sprint and is going to be available in CIS 2.2.1. You could try the following build amit49g/k8s-bigip-ctlr:as324

- Create ConfigMap to support ports below. 

```
spec:
  ports:
    - name: http
      protocol: TCP
      port: 80 -------- CIS is only seeing this. 
      targetPort: 80
      nodePort: 31991
    - name: https
      protocol: TCP
      port: 443 ---- CIS is ignoring this. 
      targetPort: 443
      nodePort: 31130
```
You need the following ConfigMap

```
kind: ConfigMap
apiVersion: v1
metadata:
  name: as3-tcp-config
  namespace: nginx-ingress
  labels:
    f5type: virtual-server
    as3: "true"
data:
  template: |
    {
        "class": "AS3",
        "declaration": {
            "class": "ADC",
            "schemaVersion": "3.20.0",
            "id": "chimera-service-tcp",
            "Nginx_IC": {
                "class": "Tenant",
                "Nginx_IC_vs_80": {
                    "class": "Application",
                    "template": "generic",
                    "VS_tcp_80": {
                        "class": "Service_TCP",
                        "virtualAddresses": [
                            "35.239.225.75"
                        ],
                        "iRules": [
                            {
                                "bigip": "/Common/Proxy_Protocol_iRule"
                            }
                        ],
                        "shareAddresses": true,
                        "virtualPort": 80,
                        "pool": "Nginx_IC_pool_80"
                    },
                    "Nginx_IC_pool_80": {
                        "class": "Pool",
                        "monitors": [
                            "tcp"
                        ],
                        "members": [
                            {
                                "servicePort": 80,
                                "serverAddresses": []
                            }
                        ]
                    }
                },
                "Nginx_IC_vs_443": {
                    "class": "Application",
                    "template": "generic",
                    "VS_tcp_443": {
                        "class": "Service_TCP",
                        "virtualAddresses": [
                            "35.239.225.75"
                        ],
                        "iRules": [
                            {
                                "bigip": "/Common/Proxy_Protocol_iRule"
                            }
                        ],
                        "shareAddresses": true,
                        "virtualPort": 443,
                        "pool": "Nginx_IC_pool_443"
                    },
                    "Nginx_IC_pool_443": {
                        "class": "Pool",
                        "monitors": [
                            "tcp"
                        ],
                        "members": [
                            {
                                "servicePort": 443,
                                "serverAddresses": []
                            }
                        ]
                    }
                }
            }
        }
    }
```

2) Another option is move to CRDs. Its super simply. I can help you get that configured if you want.

**Comment by mdditt2000** [comment]
@nergalex if you want to try NCC change the CIS arguments

```
 args: 
            - "--bigip-username=$(BIGIP_USERNAME)"
            - "--bigip-password=$(BIGIP_PASSWORD)"
            - "--bigip-url=https://10.128.0.38:8443"
            - "--bigip-partition=chimera"
            - "--namespace=nginx-ingress"
            - "--pool-member-type=cluster"
            - "--flannel-name=fl-vxlan"
            - "--log-level=DEBUG"
            - "--insecure=true"
            - "--log-as3-response=true"
            - "--custom-resource-mode=true"
            - "--nginx-cis-connect-mode=true"
          command: 
```
Please use above version. Please understand it's only preview

**Comment by nergalex** [comment]
Thank you for your answer.
Regarding ConfigMap solution (see attached file), pool members are still absent. 2 VS and 2 pools are well created.
Another test: I updated the Service by deleting the port structure below and still no pool members
```
    - name: http
      protocol: TCP
      port: 80
      targetPort: 80
      nodePort: 31991
```

I'll try tomorrow CRD and NCC solutions.

[ConfigMap-as3-spoke-ingress-controller-standard.yaml.TXT](https://github.com/F5Networks/k8s-bigip-ctlr/files/5586506/ConfigMap-as3-spoke-ingress-controller-standard.yaml.TXT)

**Comment by chen23** [comment]
something to check is to make sure your [configmap ](https://github.com/F5Networks/k8s-bigip-ctlr/files/5584044/ConfigMap.yaml.TXT)is in the same namespace as your NGINX Ingress Controller "spoke-ingress-controller-standard"

**Comment by nergalex** [comment]
@chen23 ConfigMap was moved to the same namespace as NGINX Ingress Controller "spoke-ingress-controller-standard".
BIG-IP object are well created again, except pool members.

**Comment by nergalex** [comment]
Thank you for your help !
Error:
- 1 port per Service supported. Fix in next CIS 2.2.1
- same namespace for CIS ConfigMap and NGINX IC Service
- ... don't forget to change label name on Service between 2 tests :op
